### ISSUE TITLE
Improper blocking of async functions 

### ISSUE THREAD

**Issue by phxnsharp** [rebus-org/Rebus.MongoDb#14]
I can say from a number of very painful experiences that Microsoft's TAP async/await model handles synchronous blocking code extremely poorly. Annoyingly, it tends to show up by deadlocking, or at least extremely poor performance, in completely unrelated code. 

This code uses a spinlock, which will block the current TAP thread. 

https://github.com/rebus-org/Rebus.MongoDb/blob/15ab99a84dac3cf2a981ce32050835a4db4306fe/Rebus.MongoDb/MongoDb/Transport/MongoDbTransport.cs#L214

The correct fix would be to use a [SemaphoreSlim](https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-7.0), which allows you to await access without a spinlock. As SemaphoreSlim already enables a maximum concurrent access count, the code would likely be both more correct, and simpler to read.

**Comment by mookid8000** [comment]
I've actually removed the MongoDB transport from Rebus.MongoDb 8 for now, because I found out that the transport was lacking in other ways too. The most severe problem was that it could potentially lose messages if the network connection went bad or if the process crashed while handling a message, and that makes it a no-go as a Rebus transport.

I am thinking about porting code from [my MongoDB queue project](https://github.com/rebus-org/MongolianBarbecue) to Rebus.MongoDb, but I have some other higher priorities at the moment, so I can't say when I get the time to look at it.
